### DISCUSSION TITLE
Promise appears when I try to create a detail page from data in a database

### DISCUSSION THREAD

**Discussion by JannikUekoetter** [neuefische/questions#280] [Web] [answered]
Hello dear underwater creatures,

I currently have the problem that i try to export data from my db.js and display them in my [id].js as single items. Should mean: within the [id].js the single item should be displayed via dynamic routing in next.js.

Because the data is local in my database I don't want to do this via fetch and just export the function from db.js to [id].js. 

Db.js:

```
function getItemByDatabaseId(id) {
  return lostItems.find((lostItem) => lostItem.id === id);
}

export { lostItems, foundItems, getItemByDatabaseId };
```

[id].js:

```
import { useRouter } from "next/router";
import { getItemByDatabaseId } from "../../components/Db";

function ItemDetail() {
  const router = useRouter();
  const itemdetailsId = router.query.id;

  const detailItem = getItemByDatabaseId(itemdetailsId);
  console.log(itemdetailsId);
  console.log(detailItem);
  return (
    <>
      <h1>
        Detail Page is working with dynamic IDs. This ID is: {itemdetailsId}.
      </h1>
      <p>{detailItem.id}</p>
    </>
  );
}

export default ItemDetail;
```

With a console.log of the detailItem I currently get the following error message in the dev tools: "Promised response from onMessage listener went out of scope". In addition, the desired object is displayed to me as promise: "Promise { <state>: "fulfilled", <value>: undefined }".
As I understand it, no promise should be displayed here, since I do not fetch asynchronously and export the data locally between two files.
What could be the reason for this and how can I fix it?
![Bildschirmfoto 2022-10-25 um 12 02 55](https://user-images.githubusercontent.com/101580574/197744904-fa6445e1-955e-49fa-a03c-884ec133de3a.png)

**Comment by mpagels** [comment]
Hey @JannikUekoetter happy to help, but could you post a link to your branch in your repo where the problem occurs? Thx.

**Comment by F-Kirchhoff** [comment]
From the next.js documentation:
> Pages that are statically optimized by [Automatic Static Optimization](https://nextjs.org/docs/advanced-features/automatic-static-optimization) will be hydrated without their route parameters provided, i.e query will be an empty object ({}).
> 
> After hydration, Next.js will trigger an update to your application to provide the route parameters in the query object.

This is what happens in your code, if you use "useRouter" you will always have to wait for the hydration to give you the id inside query. You can try to use a useEffect with a state for your detailItem like this:

```js
const [detailItem, setDetailItem] = useState(null);

useEffect( () => {
  const foundItem = getItemByDatabaseId(router.query.id)
  setDeteailItem(foundItem)
},[])

if(!detailItem) return <p> loading... </p>

return ...
```

**Comment by F-Kirchhoff** [comment, marked as answer]
An alternative is to use "getServerSideProps". In this way you don't need useRouter and useEffect, since the Component gets the id from the url parameter as a prop:
```js
export function getServerSideProps(context) {
  return { props: { id: context.query.id } };
}

function ItemDetail({ id }) {
  const detailItem = getItemByDatabaseId(id);
  ...
```
